bios/main.c

```diff
@@ -213,52 +193,47 @@ void main_init_sec(struct sec_entry_arg *arg);
 void main_init_sec(struct sec_entry_arg *arg)
 {
 	void *fdt;
-	int r;
-	const uint8_t *sblob_start = &__linker_secure_blob_start;
-	const uint8_t *sblob_end = &__linker_secure_blob_end;
+	long r;
 	struct optee_header hdr;
-	size_t pg_part_size;
-	uint32_t pg_part_dst;
 
 	msg_init();
 
 	/* Find DTB */
-	fdt = open_fdt(DTB_START, &__linker_nsec_dtb_start,
-			&__linker_nsec_dtb_end);
+	fdt = open_fdt(DTB_START);
 	r = fdt_pack(fdt);
 	CHECK(r < 0);
 
-	/* Look for a header first */
-	CHECK(((intptr_t)sblob_end - (intptr_t)sblob_start) <
-		(ssize_t)sizeof(hdr));
-	copy_bios_image("secure header", (uint32_t)&hdr, sblob_start,
-			sblob_start + sizeof(hdr));
-
+	r = semihosting_download_file("tee-header_v2.bin",
+				      sizeof(hdr), (uintptr_t)&hdr);
+	CHECK(r < 0);
 	CHECK(hdr.magic != OPTEE_MAGIC || hdr.version != OPTEE_VERSION);
 
 	msg("found secure header\n");
-	sblob_start += sizeof(hdr);
-	CHECK(hdr.init_load_addr_hi != 0);
-
-	pg_part_size = sblob_end - sblob_start - hdr.init_size;
-	pg_part_dst = (size_t)TZ_RES_MEM_START + TZ_RES_MEM_SIZE - pg_part_size;
+	CHECK(hdr.arch != OPTEE_ARCH_ARM32);
+	CHECK(hdr.num_images < 1 || hdr.num_images > OPTEE_MAX_NUM_IMAGES);
+	CHECK(hdr.images[0].image_id != OPTEE_IMAGE_ID_PAGER);
 
-	copy_bios_image("secure paged part",
-			pg_part_dst, sblob_start + hdr.init_size, sblob_end);
+	arg->entry = hdr.images[0].load_addr_lo;
+	arg->paged_part = 0;
 
-	sblob_end -= pg_part_size;
-
-	arg->paged_part = pg_part_dst;
-	arg->entry = hdr.init_load_addr_lo;
+	r = semihosting_download_file("tee-pager_v2.bin", 2 * 1024 * 1024,
+				      arg->entry);
+	CHECK(r < 0);
 
-	/* Copy secure image in place */
-	copy_bios_image("secure blob", hdr.init_load_addr_lo, sblob_start,
-			sblob_end);
+	if (hdr.num_images > 1) {
+		uint32_t dst = hdr.images[1].load_addr_lo;
+
+		CHECK(hdr.images[1].image_id != OPTEE_IMAGE_ID_PAGED);
+		CHECK(hdr.images[1].load_addr_hi != 0xffffffff);
```

Comment:
Playing with the format showed me an minor issue: `load_addr_hi` can be null if `dsp != 0xffffffff`.